server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java

```diff
@@ -60,131 +87,179 @@
                 }
             }
-            if (status != 1) {
-                String[] rms = successList.toArray(new String[0]);
-                if (rms.length > 0) {
-                    jedis.del(rms);
+
+            //If someone has failed,all the lockkey which has been added need to be delete.
+            if (FAILED.equals(status)) {
+                if (successSet.size() > 0) {
+                    Pipeline pipeline2 = jedis.pipelined();
+                    successSet.forEach(locKey ->
+                            pipeline2.hdel(locKey, XID, TRANSACTION_ID, BRANCH_ID, RESOURCE_ID,
+                                    TABLE_NAME, ROW_KEY, PK));
+                    pipeline2.sync();
                 }
                 return false;
-            } else {
-                try {
-                    String xidLockKey = getXidLockKey(locks.get(0).getXid());
-                    jedis.lpush(xidLockKey, readyKeys.toArray(new String[0]));
-                } catch (Exception e) {
-                    return false;
-                }
-                return true;
             }
+            String xidLockKey = buildXidLockKey(needLockXid);
+            StringJoiner lockKeysString = new StringJoiner(SEMICOLON);
+            needLockKeys.stream().forEach(lockKey -> lockKeysString.add(lockKey));
+            jedis.hsetnx(xidLockKey, branchId.toString(), lockKeysString.toString());
+            return true;
         }
     }
 
     @Override
     public boolean releaseLock(List<RowLock> rowLocks) {
         if (CollectionUtils.isEmpty(rowLocks)) {
-            // no lock
             return true;
         }
-        String[] keys = new String[rowLocks.size()];
-        List<LockDO> locks = convertToLockDO(rowLocks);
-        for (int i = 0; i < locks.size(); i++) {
-            String key = getLockKey(locks.get(i).getRowKey());
-            keys[i] = key;
-        }
+        String currentXid = rowLocks.get(0).getXid();
+        Long branchId = rowLocks.get(0).getBranchId();
+        List<String> needReleaseKeys = new ArrayList<>(rowLocks.size());
+        List<LockDO> needReleaseLocks = convertToLockDO(rowLocks);
+        needReleaseLocks.stream()
+                .forEach(lockDO -> needReleaseKeys.add(buildLockKey(lockDO.getRowKey())));
+
         try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
-            String xidLockKey = getXidLockKey(locks.get(0).getXid());
-            Pipeline pipeline = jedis.pipelined();
-            pipeline.del(keys);
-            Arrays.stream(keys).forEach(key -> pipeline.lrem(xidLockKey, 0, key));
-            pipeline.sync();
+            Pipeline pipelined = jedis.pipelined();
+            needReleaseKeys.stream()
+                    .forEach(key -> pipelined.hmget(key, XID, TABLE_NAME, PK, BRANCH_ID));
+            List<Object> existedObjs = pipelined.syncAndReturnAll();
+            List<List<String>> existedLockInfos = (List<List<String>>) (List) existedObjs;
```

Comment:
the lock can only be actively unlocked by the owner without having to retrieve it again

Reply:
finished